tensorrec/loss_graphs.py

```diff
@@ -54,59 +63,50 @@ def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
         return tf.sqrt(tf.reduce_mean(tf.square(error)))
 
 
-class SeparationLossGraph(AbstractLossGraph):
-    """
-    This loss function models the explicit positive and negative interaction predictions as normal distributions and
-    returns the probability of overlap between the two distributions.
-    Interactions can be any positive or negative values, but this loss function ignored the magnitude of the
-    interaction -- interactions are grouped in to {i < 0} and {i > 0}.
-    """
-    def loss_graph(self, tf_prediction_serial, tf_interactions_serial, **kwargs):
-
-        tf_positive_mask = tf.greater(tf_interactions_serial, 0.0)
-        tf_negative_mask = tf.less_equal(tf_interactions_serial, 0.0)
-
-        tf_positive_predictions = tf.boolean_mask(tf_prediction_serial, tf_positive_mask)
-        tf_negative_predictions = tf.boolean_mask(tf_prediction_serial, tf_negative_mask)
-
-        tf_pos_mean, tf_pos_var = tf.nn.moments(tf_positive_predictions, axes=[0])
-        tf_neg_mean, tf_neg_var = tf.nn.moments(tf_negative_predictions, axes=[0])
-
-        tf_overlap_distribution = tf.contrib.distributions.Normal(loc=(tf_neg_mean - tf_pos_mean),
-                                                                  scale=tf.sqrt(tf_neg_var + tf_pos_var))
-
-        loss = 1.0 - tf_overlap_distribution.cdf(0.0)
-        return loss
-
-
 class WMRBLossGraph(AbstractLossGraph):
     """
     Approximation of http://ceur-ws.org/Vol-1905/recsys2017_poster3.pdf
     Interactions can be any positive values, but magnitude is ignored. Negative interactions are also ignored.
     """
     is_dense = True
+    is_sample_based = True
 
-    def loss_graph(self, tf_interactions, tf_prediction, **kwargs):
+    def loss_graph(self, tf_prediction, tf_interactions, tf_sample_predictions, **kwargs):
+
+        # WMRB expects bounded predictions
+        tanh_prediction = tf.nn.sigmoid(tf_prediction)
+        tanh_sample_prediction = tf.nn.sigmoid(tf_sample_predictions)
```

Comment:
Think it is ok to hard code with sig? or maybe warn is prediction value is unbounded

Reply:
I think hard-coding is okay given that tf_prediction would be an unbounded dot product in all cases and could violate the [0, 1] that WMRB expects. It'd be tough to warn when prediction is unbounded at graph build time.